src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs

```diff
@@ -0,0 +1,108 @@
+                FetchEncodedChunk,
+                (File.Size + _maxChunkSize - 1) / _maxChunkSize,
+                Math.Max(1, maxBufferSize / _maxChunkSize)); // Degree of parallelism on fetch.
+            _currentChunkDecodingBuffer = new byte[_maxChunkSize];
+        }
+
+        protected override async ValueTask<int> CopyFileDataIntoBuffer(long sourceOffset, Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            var totalBytesCopied = 0;
+
+            while (destination.Length > 0)
+            {
+                // If we don't yet have a chunk, or it's fully consumed, get the next one.
+                if (!_currentChunk.HasValue || _currentChunkDecodingBufferConsumedLength == _currentChunk.Value.LengthBytes)
+                {
+                    // If we've already read some data, and the next chunk is still pending,
+                    // then just return now rather than awaiting.
+                    if (totalBytesCopied > 0 && _chunkSequence.TryPeekNext(out var nextChunk) && !nextChunk.Base64.IsCompleted)
+                    {
+                        break;
+                    }
+
+                    _currentChunk = _chunkSequence.ReadNext(cancellationToken);
+
+                    var currentEncodedChunk = await _currentChunk.Value.Base64;
+
+                    DecodeChunkToBuffer(currentEncodedChunk, _currentChunkDecodingBuffer, 0, _currentChunk.Value.LengthBytes);
+
+                    _currentChunkDecodingBufferConsumedLength = 0;
+                }
+
+                // How much of the current chunk can we fit into the destination?
+                var numUnconsumedBytesInChunk = _currentChunk.Value.LengthBytes - _currentChunkDecodingBufferConsumedLength;
+                var numBytesToTransfer = Math.Min(numUnconsumedBytesInChunk, destination.Length);
+
+                if (numBytesToTransfer == 0)
+                {
+                    break;
+                }
+
+                // Perform the copy.
+                new Memory<byte>(_currentChunkDecodingBuffer, _currentChunkDecodingBufferConsumedLength, numBytesToTransfer).CopyTo(destination);
+                destination = destination.Slice(numBytesToTransfer);
+                _currentChunkDecodingBufferConsumedLength += numBytesToTransfer;
+                totalBytesCopied += numBytesToTransfer;
+            }
+
+            return totalBytesCopied;
+        }
+
+        private EncodedFileChunk FetchEncodedChunk(long index, CancellationToken cancellationToken)
+        {
+            var sourceOffset = index * _maxChunkSize;
+            var chunkLength = (int)Math.Min(_maxChunkSize, File.Size - sourceOffset);
+            var task = _jsRuntime.InvokeAsync<string>(
+                InputFileInterop.ReadFileData,
+                cancellationToken,
+                _inputFileElement,
+                File.Id,
+                index * _maxChunkSize,
+                chunkLength).AsTask();
```

Comment:
This is untrusted data, we should validate that the "data" that was returned has the correct format and size we expect and abort the process as soon as some of it is not valid.

Reply:
https://github.com/dotnet/aspnetcore/blob/6aced8a9a2ad951f583500cfd114be7f2f7d3493/src/Components/Web.Extensions/src/InputFile/RemoteBrowserFileStream.cs#L59-L68